Implements Advanced Weather Framework on top of standard weather.

TerryTehCoder opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Begins the framework for a more complex weather subsystem, **including but not limited to**:

-Weather effects which can be flag assigned to each storm.
-More detailed clothing protection system which includes effect specific protection, and levels of protection.
-Map Customizable, almost all of the details about a storm such as desc, messages, durations, probabilities, etc, can be changed in a maps Json config file. 

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

I was quite inspired by Nebulas weather system, and while I hate to dirty up this beautiful codebase I have stumbled into, I quite fancy the features that it provided. Daedalus's weather is fairly similar to TG, and provides a good starting point for more advanced effects, like Wind, Lightning, Thunder, Flooding, more severe environmental damages, etc. This is something I've wanted to implement on My personal mapping project, but also something I think could benefit every Site. 

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
Added: Map-Specific storm overrides.
Added: Framework for clothing protection from individual weather effect flags
Added: Wind Gusts, Lightning Strikes, Fog, Tornados
Added: Rainstorm weather type, sprites ported from Nebula. 
Added: Chunking system for weather effects which handle atom lists, areas/global effects are handeled on their own.
\:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TerryTehCoder committed:
Initial Commit

Made the framework for weather_effects which can be assigned to individual storms. Also made storms map-specific customizable.

Framework for clothing protection system from weather effects.

Only one weather effect right now and one additional storm (Rainstorm)

Wind Gusts:
-Can push/throw people
-Can push/throw objects
-Pays attention to the weight of people/objects which is influenced by severity.
-plays sounds for gusts
-Clothing protection yatta yatta

TerryTehCoder committed:
Adds final wind gust sound.

Weather ambience sound didn't commit for some unspeakable reason.

TerryTehCoder commented:
I am very sleepy, and still working on this, it is prone to terrible chunks of code, as such I have marked it as a draft.

TerryTehCoder committed:
Fixes Compilation Errors

No code problems detected on this compile, but could be Mechanical issues that evaded my notice. No testing done yet.

TerryTehCoder committed:
Cleanup/Lightning Strikes

Cleaned up code from before, removed magic numbers, added random effect selection for storms so they'll actually randomize the effects they have (from a weighted acceptable effects list which is also map customizable).

Split weather fire into handeling for:
-Mob Specific Effects
-Object Specific Effects
-Area Specific Effects
-Global Effects independent of mobs

I hope this won't make the subsystem too malperformant; if it does it would be changeable but I think it Should be alright, I'd much rather have the option to run weather independent for these things than stuck acting on Exclusively mobs. This allows me to do things like have lighting strikes randomly across the map regardless of player presence, which can lead to cool scenarios like critical equipment becoming damaged requiring engineering intervention, a core tenant of the whole point in a weather system.

Added the lightning strikes effect framework, I THINK I'm fairly finished with wind gusts, added a mob independent (see global) version that should give us that awesome (equipment pushed around on a weather deck) vibe I'm looking for.

Though I have to be careful because obj lists are extensive and I could easily cut performance if I don't tread lightly.

Added a outdoor_weather_objects global list which obj's get put into when they're moved from an indoor area to an outdoor area, and removed viceversa. May need additional visiting for checking player contents like backpack interiors/equipment.

TerryTehCoder commented:
Note mostly for myself but:

Fire() --> apply_to_(obj, mob, area, etc) --> weather_act() --> apply_effect()

1. Filtration/Batching/Lists?

2. Allows us to avoid excessive loops

3. Verifies total immunities before passing to effects. (Can_Act is probably checking other stuff? Otherwise can we handle this somewhere else?)
[Expansion to handle additional calls from each apply_to? What would that even look like?? Ideally it would handle additional arguments for atoms that could be obj, area, mob, or other things and then it would pass that information on to the relevant effect after checking for immunities and whatever other esoteric bullshit it does???]

4. Effect actually gets applied, effect specific protections evaluated against "parent". Sharing effect information with Subsystem for filtering.

I'm really starting to wonder if I can just cut out the middleman. The batched Apply_to system is great because I can handle individual arguments much more easily, but I'm constraining just to fit it through weather_act.

Who Knows.

TerryTehCoder committed:
God knows What

I am forgetting what I did in this revision.

Built a chunking system for determining outdoor chunks, they'll probably query chunks they recieve from the coverage determination system.

Setup some dirty flag verifications. Simulated threading for cutting the weather processing into more manageable pieces in-case the entire crew decides to hold a pizza party or someone releases 10,000 pieces of glass on the weather deck, added some earlier filtering, and selective effect filtering.

Also tick interval check for cooldown.

TerryTehCoder committed:
The Turf Chunkening/Weather Profiles/Coverage

Adds chunking system for turfs, switches the initial weather telegraph to be entirely based on turfs rather than area.

My goal is to eventually move everything to turfs because I am a monster and I want to see players suffer.

Begins work on weather profiles, just a light framework for now.

Added the weather coverage subsystem which calculates exposed turfs at init/listens for turf updates and passes the information to the chunking system.

The chunking system then distributes that information to all the other systems. They basically all pool information in the chunking system because he's such a cool guy (He is also weirdly efficient for being so chonky).

TD:
Finish weather profiles, integrate with other systems.

More weather effects

Cleanup the rest of the weather code that's still stuck in the past (using areas instead of turfs).

Profiling???

ponut64 commented on code/controllers/subsystem/weather_coverage:
```diff
@@ -0,0 +1,119 @@
+				// This turf blocks weather—everything below is covered.
+				break
+
+			// If there was no turf at the top or all were null, mark the one below maxz as exposed.
```
This seems oddly specific; why wouldn't you search for turfs at every Z below the minimum, and thusly be able to register every turf below the ones which do not block turf into the weather system? In that case you get one early exit, and that is simply that a turf blocker exists here at this (x,y) and nothing else at this (x,y) need be registered; Such that a loop iterates from (minz) to (maxz) per (xy) and only needs to break if there is no turf (!T) or the turf blocks weather. Otherwise, register it as exposed.

TerryTehCoder replied:
Maps and roofing areas exist at different elevations in SS13. This system (on most maps assuming atleast 2-Z's) is likely to stop for most turfs on the Maxz. 

Something like my map has many Z's which are frequently topped by roofing areas, consequently the loop would be broken relatively soon for these areas vs going bottom up which would likely take a portion of time in comparison. 

I think I may be misunderstanding your idea. I believe you are suggesting: instead of Maxz to 1, we go 1 to Maxz, marking every turf up until a blocker as exposed, and early exiting only when finding a blocker (as opposed to a blocker/non-blocker).

This method would probably be more thorough but would probably take longer due to the bulk size of turfs. With my method we *Could* end up in the awkward situation where a turf is exposed beneath a turf that is protecting, and it has been flagged as COVERED but the signal system will ensure this is re-evaluated only when the turf above it is destroyed.

Essentially instead of caching one large list from bottom to top we loop through What I believe is more quick iteration on larger Z maps, from top to bottom, stopping once we have a single turf since at roundstart we don't Care about the lower turfs; the signals system will tell us if we need to update them or not, and dirty flags will Ideally prevent any accidents. 

Did I get your idea right? I only had 6 hours of sleep, but I feel quite fine.

TerryTehCoder replied:
Resolved as per our last discussion.

ponut64 commented on code/controllers/subsystem/weather_coverage:
```diff
@@ -0,0 +1,119 @@
+
+/datum/controller/subsystem/weather_coverage/fire()
+	for(var/turf/T in exposed_turfs)
+		if (!T.weather_dirty)
```
I don't see this referenced anywhere else?

TerryTehCoder replied:
When the weather coverage subsystem fires, it checks the exposed_turfs list from the chunking system. The only thing I forgot is that exposed_turfs should be pulled from weather_chunking field access which is probably what you're confused about. 

There is a chunking system setup for getting specific areas more efficiently but for the purposes of global iterating like this I think it's fine to just check the raw list.

TerryTehCoder replied:
Ah perhaps you were talking about weather_dirty, in which case I realized that was antiqated from my framework and replaced with needs_weather_update on atoms; I've fixed this now.

TerryTehCoder replied:
I actually don't even remember why I was firing this here. The signals system handles turf updates, and that's passed off to the chunk system which handles adding it to the list. The only thing they don't do which I need to add is updating dirty flags for the turfs.

ponut64 commented on code/controllers/subsystem/weather_coverage:
```diff
@@ -0,0 +1,119 @@
+		update_cover_cache(T)
+
+
+/datum/controller/subsystem/weather_coverage/proc/weather_update_below(var/turf/T)
```
the loop which requires this function could be easily structured to not need it

ponut64 commented on code/datums/weather/weather.dm:
```diff
@@ -228,23 +268,52 @@
-	return
+/datum/weather/proc/weather_act(mob/living/L, /obj/O, /area/A)
+	for(var/datum/weather/effect/E in weather_effects)
+		if(!istype(E, /datum/weather/effect)) //Sanity Check? AlmonD Water.
```
You might want to report if an empty or invalid type is found, as opposed to simply chuck it. Find a way to report it as immune?

TerryTehCoder replied:
Follow up: Weather_Act gutted to serve as an optional override for unique effects in individual storm types.

ponut64 commented on code/datums/weather/weather.dm:
```diff
@@ -228,23 +268,52 @@
+
+			//Adding new overlays
+			if(length(new_overlay_cache))
+				T.overlays += new_overlay-cache
```
I think this line has a typo your interpreter did not catch (the danger of using plain english declarations): new_overlay-cache as opposed to new_overlay_cache

TerryTehCoder replied:
It caught it, I just troubleshoot in 12 hour periods instead of fixing them one at a time like a reasonable person. Fixed.

ponut64 commented on code/game/objects/objs.dm:
```diff
@@ -361,3 +361,20 @@
+
+/obj/Move(new_loc, dir = 0)
+	if(!new_loc)
+		return ..(new_loc, dir)
```
i'm unfamiliar with this syntax but I can't imagine this would put you in anything but an infinite loop

ponut64 commented on code/datums/weather/weather_effects.dm:
```diff
@@ -0,0 +1,410 @@
+	RegisterSignal(COMSIG_MOVABLE_MOVED, /datum/weather/effect/proc/outdoor_atom_moved)
+
+/datum/weather/effect/proc/outdoor_atom_added(atom/movable/A)
+	if(!A || A.anchored) //How did it get moved here if it was anchored? I don't know.
```
If an anchored atomic could be unanchored later or is otherwise interactable, why exclude it?

TerryTehCoder replied:
You're right in that I should probably let the effects code handle this on a case by case basis instead of globally deciding they're not outdoors because they're anchored. 

I suppose there are also some use cases where players could manage to move anchored objects out of pure shenanigans. Updated.

ponut64 commented on code/datums/weather/weather_effects.dm:
```diff
@@ -0,0 +1,410 @@
+/datum/weather/effect/proc/apply_effect(mob/living/L, protection_flag)
+	var/protection_level = 0
+	for (var/obj/item/clothing/equipped_item in L.contents)
+		if(equipped_item.weather_protection_flags & protection_flag)
```
I don't see "protection_flag" as a single bit defined constant? Am I misreading the use of "&" (bitwise AND)

TerryTehCoder replied:
Protection_flag is an argument that child procs of apply_effect pass in to determine if clothing on the mob has protection of that specific type. 

Each effect child proc that needs it (such as one's that effect mobs) calls the parent with a mob and a protection_flag such as CLOTHING_LIGHTNINGPROOF. 

![image](https://github.com/user-attachments/assets/524f8c05-8c08-4abe-87c4-e42331e777c5)

TerryTehCoder replied:
This is a temporary example since I will probably change how that specific lightning strike proc works.

TerryTehCoder replied:
Er, my mistake. I called them children because they used to be before I changed it, but right now they're not really children. They're just manually calling apply_effect with the relevant arguments. My mistake! The logic is the same though.

ponut64 commented on code/datums/weather/weather_effects.dm:
```diff
@@ -0,0 +1,410 @@
+
+/datum/weather/effect/wind_gust/proc/apply_gust_effect(mob/living/L, direction)
+	if(src.apply_effect(L, CLOTHING_WINDPROOF)) //We take different arguments than apply_effect, so it's not our parent, and we need to call it manually.
+		return //If the player has full protection, don't apply the effect, otherwise continue.
```
I don't (personally) think the return condition should be before sounds are played.

TerryTehCoder replied:
Good catch. I suppose people DO usually hear the wind even when they are protected from it. Updated.

ponut64 commented on code/datums/weather/weather_effects.dm:
```diff
@@ -0,0 +1,410 @@
+
+	else if(strike_mode == STRIKE_NEAR_PLAYER)
+		for(var/mob/living/player in GLOB.player_list)
+			if(isturf(player.loc))
```
Empty outcome

TerryTehCoder replied:
This is redundent now as the lightning strike effect was rewritten and properly atomized to the different areas.

ponut64 commented on code/datums/weather/weather_effects.dm:
```diff
@@ -0,0 +1,410 @@
+	var/list/turfs = list()
+	var/max_attempts = 20 //If we fail to find a target in 10 attempts, we pack it up and go home.
+
+	if(!turfs || turfs.len < 10) //If there's no turfs to strike (or less than 10), we need to find some, else lets just use the one's we have.
```
am i misreading the tabs here or is this condition mistakenly written as a precondition for having a strike at all

TerryTehCoder replied:
It *Is* the conditional for a strikes Turf selection but it was written before I had access to a global exposed turf list. Good Catch. It will be revamped and hopefully made much more sensical.

TerryTehCoder replied:
We're now properly using the weather chunking systems reference for exposed turfs, and this is antiquated. Fixed.

TerryTehCoder committed:
Multiple new System implements.

I cannot even begin to describe all of these changes, but I will try.

-Adds the map config changes I was testing on Theseus, this is an example of what our configs would look like as opposed to the old bay DM style we're familiar with. Hopefully this will give you an idea of what it will be like to change the config settings. (Though I understand it does not seem intuitive)

-Adds chunking/profile system reference to the weather subsystem file, changed mob/obj candidate list to get info from chunking system. Fixed some batch errors. After removing area_types (since we're trying to move to a turf dependent system) I've redone a lot of Z_checks to make more sense. Added storm center/radius values, so theoretically a storm can have an "Eye" as it moves towards the center.

-Clarified some comments in the weather coverage system, as well as labels for categories, fixed some signal code, beautification, atomization of duplicate code (We're not doing kernel level programs, are we?).

-Fixed a lot of systems like weather datum procs not referencing the new chunking system, so they should be a bit more sensical now. Added weighting for allowed_weather_effects pulled from the weather profile system.

-Speaking of implemented a weather profile system which will suggest conditions at roundstart that the SS will pull from and make a final ruling on.

-Got rid of the duplicate code in weather_act since it's handeled by apply_to code in weather_effects and the main fire() in the SS. Weather_act will now serve as a hook for weather_types which want to implement unique effects that aren't suitable for weather_effects.

-Made severity a variable on each effect that is Influenced by a weather profiles severity but isn't necessarily identical. Fixed a lot of typos that were causing issues, made a lot of apply_to children actual children and not procs.

-Added a few starter profiles that I wanted to see, but the system could be expanded upon later.  Atlantic Storm Front, Fog Banks, Gale Surges, Diesel Rain, Heat Haze, Clear Skies, Equatorial Monsoons. Each profile has unique flavor texts that will play throughout the round in outside areas.

-Removed area_type, protected_areas, and protect_indoors from all of the weather_types since this weather system is intended to be based on turfs rather than areas.

TerryTehCoder committed:
Working State

Pushing a version with all linter errors/warnings fixed while it's on my mind, and before I add new content again.

TerryTehCoder committed:
Midnight Shenanigans

Adds flavor smell integration to the SSweather fire proc, random per player long text and short, pulled from profiles.

Cleans up some select_weather_effects logic to be better based on the profiles

Qdel for lightning strike verb. (Also lightning strike verb)

 Changed profile temp/pressure to randomized value in range of low, med, high + integration to SS.

Still being Haunted by that weird compiler issue with loadout.dm and preferences_savefile.dm which I fixed in  #15

TerryTehCoder committed:
Excuse my Language, but : ..() ?

I'm tired.

This one has way too much stuff to describe, just look through it.

Major: Weather-Z turf processing on map init rather than every Z. (If your Z can't do weather why are you bothering to scan it?) + Batching for that init as well. Something like 1500 per 10 deci, about 43 seconds for 65,000 turfs to be processed.

TerryTehCoder committed:
SAFETY BACKUP

I fixed the chunk generation, and then I broke it, and then I fixed it again. Please do not fail me.

TerryTehCoder committed:
6.2.25

Got chunking working (Okay?). Sometimes it fully covers the map, and other times there are little holes. I suspect math errors in coverage/chunking code.

TerryTehCoder commented:
Chunking is working correctly; I had to add a typecheck to make sure we were only scanning relevant Z's in the actual terrain checking loop, despite theoretically passing in only relevant-z's, so something is clearly funny there and worth investigating. 

The issue was that we were scanning Z's above the relevant Z's and turfs above that were weather proof were blocking the actual relevant Z turfs from recieving weather (Space Hotel blocking rain...), this is fixed in my most recent build which I've yet to push.

TerryTehCoder committed:
Cleans up Profiles/Probabilities

Adds some more detailed descriptions of various variables.

TerryTehCoder committed:
Allows Baking of Weather Coverage Data to file

Completely negates the need for calculations in most rounds; calculations will automatically run in the background when the server detects the hash has changed and a manual bake can be triggered.

TerryTehCoder committed:
Moves in the fieldtest map I was using for testing

Cleaning up my branch and I figure I can bring the testing map into the repo as well. It's a small abandoned house in a field with many trees and no atmosphere, mostly for testing weather effects.

TerryTehCoder committed:
Reworks Sound/Replaces Locate for Z-stacks with GetAbove and GetBelow

Thank you Kapu.

TerryTehCoder committed:
Optimizes Caching/Fixes Weather Coverage (Again)

GetBelow returns null in some cases which doesn't work well so it needed an exception.

Also optimized caching for coverage data so it's much smaller now (Cuts out repeating Z coordinate every time in favor of one nesting coords inside of Z-level which is shown once for each Z)

TerryTehCoder committed:
Cleanup

TerryTehCoder commented:
This was not ready to merge.
